tst/EnergyPlus/unit/HeatBalFiniteDiffManager.unit.cc

```diff
@@ -233,4 +237,114 @@ TEST_F(EnergyPlusFixture, HeatBalFiniteDiffManager_adjustPropertiesForPhaseChang
+     "   0.8;                     !- Visible Absorptance",
+        
+     "Material,",
+     "   F16 Acoustic tile, !- Name",
+     "   MediumSmooth, !- Roughness",
+     "   0.0191, !- Thickness{ m }",
+     "   0.06, !- Conductivity{ W / m - K }",
+     "   368, !- Density{ kg / m3 }",
+     "   590.000000000002, !- Specific Heat{ J / kg - K }",
+     "   0.9, !- Thermal Absorptance",
+     "   0.3, !- Solar Absorptance",
+     "   0.3;                     !- Visible Absorptance",
+        
+     "Material,",
+     "   M11 100mm lightweight concrete, !- Name",
+     "   MediumRough, !- Roughness",
+     "   0.1016, !- Thickness{ m }",
+     "   0.53, !- Conductivity{ W / m - K }",
+     "   1280, !- Density{ kg / m3 }",
+     "   840.000000000002, !- Specific Heat{ J / kg - K }",
+     "   0.9, !- Thermal Absorptance",
+     "   0.5, !- Solar Absorptance",
+     "   0.5;                     !- Visible Absorptance",
+     
+     "Construction,",
+     "  ExtSlabCarpet 4in ClimateZone 1 - 8, !- Name",
+     "  MAT - CC05 4 HW CONCRETE, !- Outside Layer",
+     "  CP02 CARPET PAD;         !- Layer 2",
+     "Construction,",
+     "   Interior Floor, !- Name",
+     "   F16 Acoustic tile, !- Outside Layer",
+     "   F05 Ceiling air space resistance, !- Layer 2",
+     "   M11 100mm lightweight concrete;  !- Layer 3",
+     "Construction:AirBoundary,",
+     "   Air Wall_ConstructionAirBoundary,  !- Name",
+     "   None,                    !- Air Exchange Method",
+     "   0;                       !- Simple Mixing Air Changes per Hour {1 / hr}", });
+
+
+    ASSERT_TRUE(process_idf(idf_objects));
+
+      
+    ErrorsFound = false;
+    GetMaterialData(*state, ErrorsFound); // read material data
+    EXPECT_FALSE(ErrorsFound);    // expect no errors
+
+    ErrorsFound = false;
+    GetConstructData(*state, ErrorsFound); // read construction data
+    EXPECT_FALSE(ErrorsFound);     // expect no errors
+
+    // allocate properties for construction objects when it is used or not for building surfaces in the model
+
+    state->dataConstruction->Construct(1).IsUsed=false;
+    state->dataConstruction->Construct(2).IsUsed = true;
+    state->dataConstruction->Construct(3).IsUsed = true;
+
+    ConstructFD.allocate(1);
+    
+    // verify whether air wall or construction not in use was excluded from finite difference method.
+    for (int ConstrNum = 1; ConstrNum <= 3; ++ConstrNum) {
+
+        if (state->dataConstruction->Construct(ConstrNum).TypeIsWindow) continue;
+        if (state->dataConstruction->Construct(ConstrNum).TypeIsIRT) continue;
+        if (state->dataConstruction->Construct(ConstrNum).TypeIsAirBoundary) continue;
+        if (!state->dataConstruction->Construct(ConstrNum).IsUsed) continue;
+        thisConstructNum = ConstrNum;
+        ConstructFD(1).Name.allocate(state->dataConstruction->Construct(ConstrNum).TotLayers);
+        thisTotalLayers = state->dataConstruction->Construct(thisConstructNum).TotLayers;
+
+     }
```

Comment:
What is happening here?  Where are you calling the E+ code to exercise your changes?  It _almost_ looks like you copied the code from inside E+ and put it in here.  If that were the case, you'd need to modify this to just call the E+ code instead.  If you have trouble calling the E+ code because it requires too much data to be set up prior to calling the function, then just put some of the modified logic into a smaller function that is easier to call.

Reply:
Thank you for your comments. I agree that directly calling the function for unit test is a better choice. The function under modification is ReportFiniteDiffInits(EnergyPlusData &state) for initialization of the finite difference calculation of each construction. I cannot figure out a way to directly call this function. So I put part of the code from the original function to run the test. Please feel free to suggest methods of call this function or modification.

Reply:
Take these four lines that are used to determine if a layer should be included or not:

![image](https://user-images.githubusercontent.com/849824/111367068-53749b80-8662-11eb-8e53-430b02f1d40c.png)

- Create a small method on the construction object named "shouldSkipForFiniteDifference".  It performs the same series of checks and returns true if the construction should be skipped.
- In the finite difference manager, replace all those lines with a call to that method on the construction object: 
```
if (state.dataConstruction->Construct(ThisNum).shouldSkipForFiniteDifference()) continue;
```
- Then your unit test can really just set up a construction object, no input objects required, and call that tiny little worker function under a series of cases and see if it returns true or not.

Done.

Reply:
Edwin, after further looking into this, I was able to call and test the original function InitialInitHeatBalFiniteDiff. Please see the updated unit test. Thank you for your comments.